# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/cairo/cairo/src/cairo-type3-glyph-surface.c
# Commit: 3cdfb39376b7
# Full Hash: 3cdfb39376b736597fe658a322389117686b7e82
# Author: Jonathan Kew <jkew@mozilla.com>
# Date: 2024-05-07 09:13:07
# Regressor Bug: 1892913
# File Overlap Count: 3
# Description:
#   Bug 1892913 - patch 1 - Update cairo source to 1.18.0 r=gfx-reviewers,lsalzman
#   
#   Wholesale replacement of files in gfx/cairo/cairo with contents of the 1.18.0 release tarball,
#   omitting subdirectories that are not part of the core library build.
#   
# ==============================================================================

diff -r 9daa351f875b -r 3cdfb39376b7 gfx/cairo/cairo/src/cairo-type3-glyph-surface.c
--- a/gfx/cairo/cairo/src/cairo-type3-glyph-surface.c	Mon May 06 19:22:35 2024 +0000
+++ b/gfx/cairo/cairo/src/cairo-type3-glyph-surface.c	Mon May 06 19:25:12 2024 +0000
@@ -182,7 +182,9 @@
 {
     cairo_type3_glyph_surface_t *surface = abstract_surface;
 
-    return _cairo_pdf_operators_fini (&surface->pdf_operators);
+    cairo_status_t status = _cairo_pdf_operators_fini (&surface->pdf_operators);
+    _cairo_surface_clipper_reset (&surface->clipper);
+    return status;
 }
 
 static cairo_int_status_t
@@ -205,6 +207,9 @@
 	return status;
 
     pattern = (const cairo_surface_pattern_t *) source;
+    if (pattern->surface->type == CAIRO_SURFACE_TYPE_RECORDING)
+	return CAIRO_INT_STATUS_IMAGE_FALLBACK;
+
     status = _cairo_surface_acquire_source_image (pattern->surface,
 						  &image, &image_extra);
     if (unlikely (status))
@@ -289,33 +294,7 @@
 					cairo_scaled_font_t  *scaled_font,
 					const cairo_clip_t     *clip)
 {
-    cairo_type3_glyph_surface_t *surface = abstract_surface;
-    cairo_int_status_t status;
-    cairo_scaled_font_t *font;
-    cairo_matrix_t new_ctm;
-
-    status = _cairo_surface_clipper_set_clip (&surface->clipper, clip);
-    if (unlikely (status))
-	return status;
-
-    cairo_matrix_multiply (&new_ctm, &surface->cairo_to_pdf, &scaled_font->ctm);
-    font = cairo_scaled_font_create (scaled_font->font_face,
-				     &scaled_font->font_matrix,
-				     &new_ctm,
-				     &scaled_font->options);
-    if (unlikely (font->status))
-	return font->status;
-
-    status = _cairo_pdf_operators_show_text_glyphs (&surface->pdf_operators,
-						    NULL, 0,
-						    glyphs, num_glyphs,
-						    NULL, 0,
-						    FALSE,
-						    font);
-
-    cairo_scaled_font_destroy (font);
-
-    return status;
+    return CAIRO_INT_STATUS_IMAGE_FALLBACK;
 }
 
 static const cairo_surface_backend_t cairo_type3_glyph_surface_backend = {
@@ -373,6 +352,7 @@
 					 glyph_index,
 					 CAIRO_SCALED_GLYPH_INFO_METRICS |
 					 CAIRO_SCALED_GLYPH_INFO_SURFACE,
+					 NULL, /* foreground color */
 					 &scaled_glyph);
     if (unlikely (status))
 	return status;
@@ -428,6 +408,7 @@
     status = _cairo_scaled_glyph_lookup (surface->scaled_font,
 					 glyph_index,
 					 CAIRO_SCALED_GLYPH_INFO_RECORDING_SURFACE,
+					 NULL, /* foreground color */
 					 &scaled_glyph);
 
     if (_cairo_int_status_is_error (status))
@@ -476,15 +457,27 @@
     _cairo_type3_glyph_surface_set_stream (surface, stream);
 
     _cairo_scaled_font_freeze_cache (surface->scaled_font);
+
+    /* Check if this is a color glyph and bail out if it is */
     status = _cairo_scaled_glyph_lookup (surface->scaled_font,
 					 glyph_index,
-					 CAIRO_SCALED_GLYPH_INFO_METRICS |
-					 CAIRO_SCALED_GLYPH_INFO_RECORDING_SURFACE,
+					 CAIRO_SCALED_GLYPH_INFO_COLOR_SURFACE,
+					 NULL, /* foreground color */
 					 &scaled_glyph);
+    if (status == CAIRO_INT_STATUS_SUCCESS) {
+	status = CAIRO_INT_STATUS_UNSUPPORTED;
+	goto FAIL;
+    } status = _cairo_scaled_glyph_lookup (surface->scaled_font,
+					   glyph_index,
+					   CAIRO_SCALED_GLYPH_INFO_METRICS |
+					   CAIRO_SCALED_GLYPH_INFO_RECORDING_SURFACE,
+					   NULL, /* foreground color */
+					   &scaled_glyph);
     if (status == CAIRO_INT_STATUS_UNSUPPORTED) {
 	status = _cairo_scaled_glyph_lookup (surface->scaled_font,
 					     glyph_index,
 					     CAIRO_SCALED_GLYPH_INFO_METRICS,
+					     NULL, /* foreground color */
 					     &scaled_glyph);
 	if (status == CAIRO_INT_STATUS_SUCCESS)
 	    status = CAIRO_INT_STATUS_IMAGE_FALLBACK;